pkg/library/home/persistentHome.go

```diff
@@ -95,3 +99,92 @@ func storageStrategySupportsPersistentHome(workspace *common.DevWorkspaceWithCon
 	storageClass := workspace.Spec.Template.Attributes.GetString(constants.DevWorkspaceStorageTypeAttribute, nil)
 	return storageClass != constants.EphemeralStorageClassType
 }
+
+func addInitContainer(dwTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec, initContainerFromConfig *v1alpha2.Container) {
+
+	initContainerComponent := v1alpha2.Component{
+		Name: constants.HomeInitComponentName,
+	}
+
+	initContainer := initContainerFromConfig
+
+	if initContainer == nil {
+		initContainer = inferInitContainer(dwTemplateSpec)
+	}
+
+	if initContainer == nil {
+		return
+	}
+
+	initContainerComponent.ComponentUnion = v1alpha2.ComponentUnion{
+		Container: &v1alpha2.ContainerComponent{
+			Container: *initContainer,
+		},
+	}
+
+	dwTemplateSpec.Components = append(dwTemplateSpec.Components, initContainerComponent)
+	dwTemplateSpec.Commands = append(dwTemplateSpec.Commands, v1alpha2.Command{
+		Id: constants.HomeInitEventId,
+		CommandUnion: v1alpha2.CommandUnion{
+			Apply: &v1alpha2.ApplyCommand{
+				Component: constants.HomeInitComponentName,
+			},
+		},
+	})
+	dwTemplateSpec.Events.PreStart = append(dwTemplateSpec.Events.PreStart, constants.HomeInitEventId)
+}
+
+func inferInitContainer(dwTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec) *v1alpha2.Container {
+	var nonImportedComponent v1alpha2.Component
+	for _, component := range dwTemplateSpec.Components {
+		if component.Container == nil {
+			continue
+		}
+
+		pluginSource := component.Attributes.GetString(constants.PluginSourceAttribute, nil)
+		if pluginSource == "" || pluginSource == "parent" {
+			// First non-imported container component is selected
+			nonImportedComponent = component
+			break
+		}
+	}
+
+	initScript := `#!/bin/sh
+
+if [ -n "$HOME_SETUP_SCRIPT" ] && [ -f "$HOME_SETUP_SCRIPT" ]; then
+	source "$HOME_SETUP_SCRIPT"
+	exit 0
+else
+	echo 'No home setup script found at $HOME_SETUP_SCRIPT'
+fi
+
+echo "Checking for stow command"
+STOW_COMPLETE=/home/user/.stow_completed
+if command -v stow &> /dev/null; then
+
+	if  [ ! -f $STOW_COMPLETE ]; then
+		echo "Running stow command"
+		stow . -t /home/user/ -d /home/tooling/ --no-folding --adopt -v 2 > /tmp/stow.log 2>&1
```

Comment:
`--adopt` should still be removed

Reply:
You're right, I'll remove it,  I had `--adopt` because without it, stow would fail when we create a conflicting file using a mounted configmap or secret like `/home/user/.config/containers/storage.conf`, for example:  https://eclipse.dev/che/docs/stable/end-user-guide/enabling-overlay-with-a-configmap/

I'm wondering, have we investigated how https://eclipse.dev/che/docs/stable/end-user-guide/requesting-persistent-storage-for-workspaces/#requesting-persistent-storage-in-a-pvc works with persistent-home? If there is a user-defined PVC with a mount path in the home directory, and if that PVC already has some data in it, I think stow could fail if there are conflicts with the contents of /home/tooling. I would have to verify though, but I would expect this to be the case.

A possible fix can be, the initcontainer could maybe read mount paths from volumes/secrets/configmap that mount under /home/user, and add them to the `.stow-local-ignore` file before running stow

Reply:
I can create a separate issue for this case

Reply:
> You're right, I'll remove it, I had `--adopt` because without it, stow would fail when we create a conflicting file using a mounted configmap or secret like `/home/user/.config/containers/storage.conf`, for example: [eclipse.dev/che/docs/stable/end-user-guide/enabling-overlay-with-a-configmap](https://eclipse.dev/che/docs/stable/end-user-guide/enabling-overlay-with-a-configmap/)

It's worth seeing if you can mount a custom `storage.conf` to `/home/tooling/.config/containers/storage.conf` instead of `/home/user/...` with a configmap and let stow create the symbolic link in $HOME. It's possible that this won't work due to the permissions that OpenShift uses when mounting files.

If this does work, then we'll have to document this quirk. 


> I'm wondering, have we investigated how [eclipse.dev/che/docs/stable/end-user-guide/requesting-persistent-storage-for-workspaces/#requesting-persistent-storage-in-a-pvc](https://eclipse.dev/che/docs/stable/end-user-guide/requesting-persistent-storage-for-workspaces/#requesting-persistent-storage-in-a-pvc) works with persistent-home? If there is a user-defined PVC with a mount path in the home directory, and if that PVC already has some data in it, I think stow could fail if there are conflicts with the contents of /home/tooling. I would have to verify though, but I would expect this to be the case.

I also think that stow would fail in this case but haven't verified it myself. There's an underlying assumption that the PVC's `/home/user/` directory would be empty. 


> A possible fix can be, the initcontainer could maybe read mount paths from volumes/secrets/configmap that mount under /home/user, and add them to the `.stow-local-ignore` file before running stow

The .stow-local-ignore actually [ignores source files](https://www.gnu.org/software/stow/manual/stow.html#Ignore-Lists) that we do not want to create symbolic links for. For example, we're ignore the `.vimrc` and cp'ing it in the UDI's entrypoint since vim does not allow the `.vimrc` to be a symbolic link for security reasons.

Home persistence definitely complicates a lot of things.  If files that are auto-mounted by OpenShift can be stowed (i.e. symbolic links for these files can be created by stow) then a potential workaround is to mount them to `/home/tooling/` in addition to `/home/user/`. I believe stow would ignore these files (since they exist in both the source and target directory), and they would remain regular files rather than symbolic links.

We'll have to investigate the case of automount volumes/secrets/configmaps conflicts with stow and potential workarounds.

Reply:
> I can create a separate issue for this case

That'd be great when you have a moment :) Thank you!!

Reply:
I created the issue: https://github.com/devfile/devworkspace-operator/issues/1257

> The .stow-local-ignore actually [ignores source files](https://www.gnu.org/software/stow/manual/stow.html#Ignore-Lists) that we do not want to create symbolic links for.

Yes, this would still help us prevent conflicts since stow won't try to create a potentially conflicting link

Reply:
^ Unless I'm misunderstanding something 


Let's continue the discussion at https://github.com/devfile/devworkspace-operator/issues/1257, I think it's out of scope of this PR

Reply:
Thanks for creating the issue :pray: 

> Yes, this would still help us prevent conflicts since stow won't try to create a potentially conflicting link

I see what you mean now (thinking of the `storage.conf` case where the same file with different contents are present in both `/home/tooling/` and `/home/user/`).

I think your solution should work as long as we ignore individual files rather than entire directories, as this might cause other files we want to create symbolic links for to be ignored by stow.